### PR TITLE
Update Guzzle to 6.3

### PR DESCRIPTION
**Pull request by phil-davis**
## Description
Update Guzzle to 6.3, and also other libraries (deleted and regenerated composer.lock)

## Motivation and Context
Guzzle 5.3 is outdated, needs update.
Also some libs used by apps will need Guzzle > 6.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

This is a newer version of #29211 made by cherry-pick and reorganize of the commits from there, to massage it all into these "logical" commits.

### CODE REVIEW

**Comment by phil-davis** [comment]
drone times out after 3 hours!!!

**Comment by phil-davis** [comment]
Lots of other bumping of versions of stuff in composer is happening just now with the dependency bot. So I will leave this conflict to resolve in a week or so when all other ``composer.lock`` changes are in master.

**Comment by phil-davis** [comment]
Last rebase 27 Dec 2017

**Comment by phil-davis** [comment]
@PVince81 how do we proceed with this Guzzle update for master/10.1?

**Comment by PVince81** [comment]
I suggest we first merge https://github.com/owncloud/core/pull/28043 because at this point all apps will become "incompatible". Then for each app we iteratively make compatible, we can think about whether said app also requires a Guzzle adjustment. If yes, then need to make it work.

At long as 10.1 is in alpha/beta, I think it's ok if some apps are not compatible yet. Once 10.1 reaches RC, we should have all common apps made compatible.

This also means that starting with this point, developing stuff on master doesn't guarantee an easy backport to stable10. But that ok I think, we need to do the jump.

Any objections to this ? @DeepDiver1975 @phil-davis @patrickjahns @felixboehm

**Comment by PVince81** [comment]
one challenge is that some storage apps will need branches: one compatible with 10.0 LTS and one for master due to Guzzle API differences
as long as 10.1 is not out I suggest that QA only focus on the ones compatible with 10.0 in the future

**Comment by phil-davis** [comment]
Rebased on top of ``master`` that has PHP 7.1 10.1prealpha.

@PVince81 @DeepDiver1975 When CI passes, this is ready.

**Comment by phil-davis** [comment]
This could wait a few days until all the app CI has been sorted out related to the core master 10.1prealpha in PHP 7.1 bumps. That way we start with clean CI in all apps against core master, and know that the next round of refactoring will be just related to these Guzzle changes.

**Comment by phil-davis** [comment]
Now that unit test warnings result in fail, these need fixing:
```
There were 5 warnings:

1) Test\Http\Client\ClientTest::testGet
Trying to configure method "get" which cannot be configured because it does not exist, has not been specified, is final, or is static

2) Test\Http\Client\ClientTest::testPost
Trying to configure method "post" which cannot be configured because it does not exist, has not been specified, is final, or is static

3) Test\Http\Client\ClientTest::testPut
Trying to configure method "put" which cannot be configured because it does not exist, has not been specified, is final, or is static

4) Test\Http\Client\ClientTest::testDelete
Trying to configure method "delete" which cannot be configured because it does not exist, has not been specified, is final, or is static

5) Test\Http\Client\ClientTest::testOptions
Trying to configure method "options" which cannot be configured because it does not exist, has not been specified, is final, or is static
```

**Comment by phil-davis** [comment]
Note: currently ``composer.lock`` here contains bumps for symfony 3.4.6 and sabre/vobject 4.1.5. That stuff needs adjusting and makes more tests fail...

**Comment by phil-davis** [comment]
Note to self: ``composer update`` found a bunch of other version bumps of dbal, symfony, zend-framework, random_compat that have been released in April. So either it all "passes easily" or I will need to make a less comprehensive ``composer.lock``

**Comment by phil-davis** [comment]
Note: I will rebase and sort this out late Friday.

**Comment by phil-davis** [comment]
Rebased. ToDo:
- [ ] look into bonus unit test fails https://drone.owncloud.com/owncloud/core/8801/158

**Comment by PVince81** [comment]
@phil-davis maybe we first need to make phan happy:
```
lib/private/Http/Client/Response.php:56 PhanUndeclaredClassMethod Call to method getBody from undeclared class \OC\Http\Client\GuzzleResponse
```

**Review by DeepDiver1975 — COMMENTED** [review]
left some minor comments ....

File: tests/lib/Http/Client/ClientTest.php [hunk 1]
```diff
@@ -26,8 +29,12 @@ class ClientTest extends \Test\TestCase {
 	public function setUp() {
 		parent::setUp();
 		$this->config = $this->createMock('\OCP\IConfig');
+		$this->response = $this->createMock('\Psr\Http\Message\ResponseInterface');
```
Comment (phil-davis) [inline tests/lib/Http/Client/ClientTest.php]: @DeepDiver1975 I was "hacking around" here sorting out what happened in Guzzle6 to the client and responses...
1) Is this the right sort of thing to do for the unit tests here?
2) Should this also have ``::class``? or it's in interface?

File: tests/Settings/Controller/CheckSetupControllerTest.php [hunk 2]
```diff
@@ -454,7 +454,7 @@ public function testIsBuggyNss400() {
 		/** @var ClientException | \PHPUnit_Framework_MockObject_MockObject $exception */
 		$exception = $this->getMockBuilder('\GuzzleHttp\Exception\ClientException')
 			->disableOriginalConstructor()->getMock();
-		$response = $this->getMockBuilder('\GuzzleHttp\Message\ResponseInterface')
+		$response = $this->getMockBuilder('\Psr\Http\Message\ResponseInterface')
```
[the remark this answers is not in the document]
Comment (DeepDiver1975) [reply tests/Settings/Controller/CheckSetupControllerTest.php]: anywhere a class name is referenced please use ::class

Usually I do this when touching any test case - and then for all occurrences.
Please have a look at https://plugins.jetbrains.com/plugin/7622-php-inspections-ea-extended- - help with this on file level.

File: tests/lib/Http/Client/ClientTest.php (hunk 1 above)
Comment (DeepDiver1975) [reply tests/lib/Http/Client/ClientTest.php]: -  createMock looks good
- and yes - use ::class here as well

File: tests/acceptance/features/bootstrap/CalDavContext.php [hunk 3]
```diff
@@ -191,16 +192,17 @@ public function userHasCreatedACalendarNamed($user, $name) {
 		$davUrl = $this->featureContext->getBaseUrl()
 			. '/remote.php/dav/calendars/' . $user . '/' . $name;
```
[the remark this answers is not in the document]
Comment (phil-davis) [reply tests/acceptance/features/bootstrap/CalDavContext.php]: This was not introduced in the PR.
Refactoring of string concatenation for acceptance test code is in PR https://github.com/owncloud/core/pull/32336

**Comment by phil-davis** [comment]
@DeepDiver1975 I have fixed up what I can.
https://drone.owncloud.com/owncloud/core/9460/158
The drone test job on objectstore clones the existing ``files_primary_s3`` ``master`` and that, of course, has Guzzle5 in it. The job gets some errors in the unit tests.

Maybe next we need to make a PR/branch in ``files_primary_s3`` that also has the Guzzle6 changes, and run CI here in ``core`` with that ``files_primary_s3`` branch?

**Comment by DeepDiver1975** [comment]
all apps which use guzzle need adoptions .... files-primary_s3, google, onedrive .......

**Comment by phil-davis** [comment]
I inherited this PR for some reason lost in the mists of time.
How do we proceed now with resource allocation and priority of getting all the needed app changes ready and planning implementation?

**Comment by DeepDiver1975** [comment]
> How do we proceed now with resource allocation and priority of getting all the needed app changes ready and planning implementation?

I suggest we plan this for the next server sprint - which starts next week. Vincent will be back by then as well.

**Comment by phil-davis** [comment]
See PR #32537 for the next iteration.

That is the Guzzle V6 code rebased and refactored for the underlying refactoring of acceptance test code that happened today. That introduced ``HttpRequestHelper`` which vastly reduced the places in acceptance test code where Guzzle client code is use. And thus it also vastly reduced the code diffs for upgrading to Guzzzle V6, which is "a god thing" (TM).

**Pull request closed without merging**